src/index.rs

```diff
@@ -1041,6 +1041,33 @@ impl Index {
             .unwrap_or_else(Vec::new)
     }
 
+    /// Returns all member variables of the given container excluding variables of type [`VariableType::Temp`]
+    /// and [`VariableType::Return`].
+    /// You should always prefer to use the [`Index::get_container_members`] method but for rare occasions
+    /// this method is needed, especially in the codegen. Specifically this method was introduced because for
+    /// ```ST
+    /// FUNCTION FOO : FOO
+    /// END_FUNCTION
+    ///
+    /// TYPE FOO : STRUCT
+    ///     bar : DINT;
+    /// END_TYPE
+    /// ```
+    /// [`Index::get_container_members`] returns `["foo", "bar"]` as its member variables because it can not
+    /// distinguish between the function `FOO`, which contains struct `foo` as a member variable, and the
+    /// struct `FOO`, which contains DINT `bar` as a member variable, merging both into one. In the codegen
+    /// this would result into a infinite-loop within the [`DataTypeGenerator::generate_initial_value`] function
+    /// yielding a stackoverflow error.
+    /// This described behaviour also becomes an issue in the [`recursive_validator::RecursiveValidator`]
+    /// function, generating incorrect warnings.
+    /// TODO: ^ is there better solution for this
+    pub fn get_container_members_filtered(&self, container_name: &str) -> Vec<&VariableIndexEntry> {
+        self.get_container_members(container_name)
+            .into_iter()
+            .filter(|it| !it.is_temp() && !it.is_return())
```

Comment:
I think skipping the return here avoids the problem but does not solve the real issue.
The problem as I see it at the moment is that during the CodeGen when asked for the datatype of the return variable we received the defined function as a type instead of receiving the actual defined type, this is because we treated the `FUNCTION` as a possible assignable type which it is not.  

The following example still fails: 
```
TYPE TM : STRUCT hours,minutes,seconds : DINT; END_STRUCT END_TYPE

FUNCTION TM : DINT
VAR_INPUT
	x  : TM;
END_VAR
END_FUNCTION
```

Let's pair on this today or on friday